### PR TITLE
[Merged by Bors] - Remove App::add_sub_app

### LINKED ISSUE #7286: Remove add_sub_app api
Consider removing the `add_sub_app` api. This method is somewhat redundant with the `insert_sub_app api`. Consider:

```rust
let sub_app = App::empty();
app.add_sub_app(SubAppLabel, sub_app, extract_function);
```

vs

```rust
let sub_app = App::empty();
app.insert_sub_app(
  SubAppLabel,
  SubApp {
    app: sub_app,
    extract: Box::new(extract_function),
  }
);
```

This is a somewhat niche feature and may not deserve an extra api for slight better ergonomics. 

See this comment from cart from the [pipelined rendering pr](https://github.com/bevyengine/bevy/pull/6503#discussion_r1081917428)

> Maybe we should just remove this in favor of `insert_sub_app`. The `insert` semantics are "correct" anyway whereas `add` is generally used for repeatable / non-overwriting actions. This function is _slightly_ more ergonomic, but SubApps aren't exactly a common pattern anyway.

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Fixes #7286. Both `App::add_sub_app` and `App::insert_sub_app` are rather redundant. Before 0.10 is shipped, one of them should be removed.

## Solution
Remove `App::add_sub_app` to prefer `App::insert_sub_app`.

Also hid away `SubApp::extract` since that can be a footgun if someone mutates it for whatever reason. Willing to revert this change if there are objections.

Perhaps we should make `SubApp: Deref<Target=App>`? Might change if we decide to move `!Send` resources into it.

---

## Changelog
Added: `SubApp::new`
Removed: `App::add_sub_app`

## Migration Guide
`App::add_sub_app` has been removed in favor of `App::insert_sub_app`. Use `SubApp::new` and insert it via `App::add_sub_app`

Old:

```rust
let mut sub_app = App::new()
// Build subapp here
app.add_sub_app(MySubAppLabel, sub_app);
```

New:

```rust
let mut sub_app = App::new()
// Build subapp here
app.insert_sub_app(MySubAppLabel, SubApp::new(sub_app, extract_fn));
```

### CODE REVIEW

File: crates/bevy_app/src/app.rs [hunk 1]
```diff
@@ -1081,28 +1089,6 @@ impl App {
         self
     }
 
-    /// Adds an [`App`] as a child of the current one.
-    ///
-    /// The provided function `extract` is normally called by the [`update`](Self::update) method.
-    /// After extract is called, the [`Schedule`] of the sub app is run. The [`World`]
-    /// parameter represents the main app world, while the [`App`] parameter is just a mutable
-    /// reference to the `SubApp` itself.
```
Comment (hymm) [inline crates/bevy_app/src/app.rs:1089]: The description of the extract function should probably be moved somewhere else. Maybe to the `SubApp` docs?

Comment (james7132) [reply crates/bevy_app/src/app.rs:1089]: Done.

File: crates/bevy_app/src/app.rs [hunk 2]
```diff
@@ -135,10 +135,23 @@ pub struct SubApp {
...
     /// A function that allows access to both the [`SubApp`] [`World`] and the main [`App`]. This is
     /// useful for moving data between the sub app and the main app.
-    pub extract: Box<dyn Fn(&mut World, &mut App) + Send>,
+    extract: Box<dyn Fn(&mut World, &mut App) + Send>,
 }
 
 impl SubApp {
+    /// Creates a new [`SubApp`].
+    ///
+    /// The provided function `extract` is normally called by the [`update`](Self::update) method.
```
Comment (hymm) [inline crates/bevy_app/src/app.rs:144]:
```suggestion
    /// The provided function `extract` is normally called by the [`update`](App::update) method.
```

**Review by hymm — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Review by mockersf — APPROVED** [review]

**Pull request closed without merging**